src/access/AccessRestrictionProvider.h

```diff
@@ -0,0 +1,225 @@
+     *
+     * If id is not set, the restriction applies to all attributes, commands, or events of the given type (wildcard).
+     */
+    struct Restriction
+    {
+        Type restrictionType;
+        Optional<uint32_t> id;
+    };
+
+    /**
+     * Defines a single entry in the access restriction list, which contains a list of restrictions
+     * for a cluster on an endpoint.
+     */
+    struct Entry
+    {
+        FabricIndex fabricIndex;
+        EndpointId endpointNumber;
+        ClusterId clusterId;
+        std::vector<Restriction> restrictions;
+    };
+
+    /**
+     * Used to notify of changes in the access restriction list and active reviews.
+     */
+    class Listener
+    {
+    public:
+        virtual ~Listener() = default;
+
+        /**
+         * Notifies of a change in the commissioning access restriction list.
+         */
+        virtual void CommissioningRestrictionListChanged() = 0;
+
+        /**
+         * Notifies of a change in the access restriction list.
+         *
+         * @param [in] fabricIndex  The index of the fabric in which the list has changed.
+         */
+        virtual void RestrictionListChanged(FabricIndex fabricIndex) = 0;
+
+        /**
+         * Notifies of an update to an active review with instructions and an optional redirect URL.
+         *
+         * @param [in] fabricIndex  The index of the fabric in which the entry has changed.
+         * @param [in] token        The token of the review being updated (obtained from ReviewFabricRestrictionsResponse)
+         * @param [in] instruction  The instructions to be displayed to the user.
+         * @param [in] redirectUrl  An optional URL to redirect the user to for more information.  May be null.
+         */
+        virtual void OnFabricRestrictionReviewUpdate(FabricIndex fabricIndex, uint64_t token, const char * instruction,
+                                                     const char * redirectUrl) = 0;
+
+    private:
+        Listener * mNext = nullptr;
+
+        friend class AccessRestrictionProvider;
+    };
+
+    AccessRestrictionProvider()          = default;
+    virtual ~AccessRestrictionProvider() = default;
+
+    AccessRestrictionProvider(const AccessRestrictionProvider &)             = delete;
+    AccessRestrictionProvider & operator=(const AccessRestrictionProvider &) = delete;
+
+    /**
+     * Set the restriction entries that are to be used during commissioning when there is no accessing fabric.
+     *
+     * @param [in] entries  The entries to set.
+     */
+    CHIP_ERROR SetCommissioningEntries(const std::vector<Entry> & entries);
+
+    /**
+     * Set the restriction entries for a fabric.
+     *
+     * @param [in] fabricIndex  The index of the fabric for which to create entries.
+     * @param [in] entries  The entries to set for the fabric.
+     */
+    CHIP_ERROR SetEntries(const FabricIndex, const std::vector<Entry> & entries);
```

Comment:
I don't see where entries are cleared on fabric removal.  That needs to happen, no?

Reply:
That is up to the app, which is responsible for managing the entries.

Reply:
We know the entries have to be removed when the fabric is removed.  Why are we forcing the app to remember to do that?

Reply:
The app is completely responsible for managing the ARL data.  It is not persisted by the SDK.  If we make the SDK also responsible for manipulating the data, there will be contention.